File: cave.java
import java.util.*;

public class Cave {

    private static final char EMPTY = '.';
    private static final char WALL = 'W';
    private static final char START = '>';
    private static final char FINISH = '<';
    private static final char COLLAPSE = 'C';

    private final int width;
    private final int height;
    private final int wallCount;
    private final int collapseCount;
    private final char[][] grid;
    private final Random random = new Random();

    public Cave(int width, int height, int wallCount, int collapseCount) {
        this.width = width;
        this.height = height;
        this.wallCount = wallCount;
        this.collapseCount = collapseCount;
        this.grid = new char[height][width];
        for (char[] row : grid) {
            Arrays.fill(row, EMPTY);
        }
    }

    public void generate() {
        placeStartFinish();
        placeWalls();
        placeCollapseSpots();
    }

    private void placeStartFinish() {
        int startY = random.nextInt(height);
        int finishY = random.nextInt(height);
        grid[startY][0] = START;
        grid[finishY][width - 1] = FINISH;
    }

    private void placeWalls() {
        for (int i = 0; i < wallCount; i++) {
            boolean horizontal = random.nextBoolean();
            int length = 1 + random.nextInt(5);
            boolean placed = false;

            for (int attempt = 0; attempt < 100 && !placed; attempt++) {
                int x = random.nextInt(width);
                int y = random.nextInt(height);
                if (canPlaceWall(x, y, length, horizontal)) {
                    for (int j = 0; j < length; j++) {
                        int nx = horizontal ? x + j : x;
                        int ny = horizontal ? y : y + j;
                        grid[ny][nx] = WALL;
                    }
                    placed = true;
                }
            }
        }
    }

    private boolean canPlaceWall(int x, int y, int length, boolean horizontal) {
        for (int i = 0; i < length; i++) {
            int nx = horizontal ? x + i : x;
            int ny = horizontal ? y : y + i;
            if (nx >= width || ny >= height || grid[ny][nx] != EMPTY) {
                return false;
            }
        }
        return true;
    }

    private void placeCollapseSpots() {
        int placed = 0;
        for (int attempts = 0; attempts < 10000 && placed < collapseCount; attempts++) {
            int x = random.nextInt(width);
            int y = random.nextInt(height);
            if (grid[y][x] == EMPTY) {
                grid[y][x] = COLLAPSE;
                applyHollowSound(x, y);
                placed++;
            }
        }
        if (placed < collapseCount) {
            throw new IllegalStateException("Not enough space to place all collapse spots");
        }
    }

    private void applyHollowSound(int cx, int cy) {
        for (int dy = -1; dy <= 1; dy++) {
            for (int dx = -1; dx <= 1; dx++) {
                int nx = cx + dx;
                int ny = cy + dy;
                if (dx == 0 && dy == 0) continue;
                if (nx >= 0 && ny >= 0 && nx < width && ny < height) {
                    char c = grid[ny][nx];
                    if (c == EMPTY) {
                        grid[ny][nx] = '1';
                    } else if (Character.isDigit(c)) {
                        if (c < '9') {
                            grid[ny][nx] = (char) (c + 1);
                        }
                    }
                }
            }
        }
    }

    public void print() {
        for (char[] row : grid) {
            for (char c : row) {
                System.out.print(c + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        if (args.length != 4) {
            System.out.println("Usage: java Cave <width> <height> <wallCount> <collapseCount>");
            return;
        }
        int width = Integer.parseInt(args[0]);
        int height = Integer.parseInt(args[1]);
        int walls = Integer.parseInt(args[2]);
        int collapses = Integer.parseInt(args[3]);

        System.out.printf("Creating a play field of size %dx%d with %d walls and %d collapsing spots\n", width, height, walls, collapses);

        Cave cave = new Cave(width, height, walls, collapses);
        cave.generate();
        cave.print();
    }
}

File: cave.py
import sys
import random

EMPTY = '.'
WALL = 'W'
START = '>'
FINISH = '<'
COLLAPSE = 'C'

class Cave:
    def __init__(self, width, height, wall_count, collapse_count):
        self.width = width
        self.height = height
        self.wall_count = wall_count
        self.collapse_count = collapse_count
        self.grid = [[EMPTY for _ in range(width)] for _ in range(height)]

    def generate(self):
        self.place_start_finish()
        self.place_walls()
        self.place_collapse_spots()

    def place_start_finish(self):
        start_y = random.randint(0, self.height - 1)
        finish_y = random.randint(0, self.height - 1)
        self.grid[start_y][0] = START
        self.grid[finish_y][self.width - 1] = FINISH

    def place_walls(self):
        for _ in range(self.wall_count):
            for _ in range(100):
                horizontal = random.choice([True, False])
                length = random.randint(1, 5)
                x = random.randint(0, self.width - 1)
                y = random.randint(0, self.height - 1)

                if self.can_place_wall(x, y, length, horizontal):
                    for i in range(length):
                        nx = x + i if horizontal else x
                        ny = y if horizontal else y + i
                        self.grid[ny][nx] = WALL
                    break

    def can_place_wall(self, x, y, length, horizontal):
        for i in range(length):
            nx = x + i if horizontal else x
            ny = y if horizontal else y + i
            if nx >= self.width or ny >= self.height or self.grid[ny][nx] != EMPTY:
                return False
        return True

    def place_collapse_spots(self):
        placed = 0
        for _ in range(10000):
            if placed >= self.collapse_count:
                break
            x = random.randint(0, self.width - 1)
            y = random.randint(0, self.height - 1)
            if self.grid[y][x] == EMPTY:
                self.grid[y][x] = COLLAPSE
                self.apply_hollow_sound(x, y)
                placed += 1
        if placed < self.collapse_count:
            raise RuntimeError("Not enough space to place all collapse spots")

    def apply_hollow_sound(self, cx, cy):
        for dy in range(-1, 2):
            for dx in range(-1, 2):
                if dx == 0 and dy == 0:
                    continue
                nx, ny = cx + dx, cy + dy
                if 0 <= nx < self.width and 0 <= ny < self.height:
                    current = self.grid[ny][nx]
                    if current == EMPTY:
                        self.grid[ny][nx] = '1'
                    elif current.isdigit():
                        if current < '9':
                            self.grid[ny][nx] = chr(ord(current) + 1)

    def print_grid(self):
        for row in self.grid:
            print(' '.join(row))


def main():
    if len(sys.argv) != 5:
        print("Usage: python cave.py <width> <height> <wallCount> <collapseCount>")
        return

    width = int(sys.argv[1])
    height = int(sys.argv[2])
    wall_count = int(sys.argv[3])
    collapse_count = int(sys.argv[4])

    print(f"Creating a play field of size {width}x{height} with {wall_count} walls and {collapse_count} collapsing spots")

    cave = Cave(width, height, wall_count, collapse_count)
    cave.generate()
    cave.print_grid()


if __name__ == "__main__":
    main()
